File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/FrequencyValueBase.vb
'*******************************************************************************************************
'  FrequencyValueBase.vb - Frequency and DfDt value base class
'  Copyright © 2008 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2008
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  11/12/2004 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports TVA.Math

''' <summary>This class represents the protocol independent a frequency and dfdt value.</summary>
<CLSCompliant(False), Serializable()> _
Public MustInherit Class FrequencyValueBase

    Inherits ChannelValueBase(Of IFrequencyDefinition)
    Implements IFrequencyValue

    Private m_frequency As Single
    Private m_dfdt As Single
    Private m_frequencyAssigned As Boolean
    Private m_dfdtAssigned As Boolean

    <Serializable()> _
    Public Enum CompositeValueType
        Frequency
        DfDt
    End Enum

    Protected Sub New()
    End Sub

    Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

        MyBase.New(info, context)

        ' Deserialize frequency value
        m_frequency = info.GetSingle("frequency")
        m_dfdt = info.GetSingle("dfdt")

        m_frequencyAssigned = True
        m_dfdtAssigned = True

    End Sub

    Protected Sub New(ByVal parent As IDataCell)

        MyBase.New(parent)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal frequency As Single, ByVal dfdt As Single)
    Protected Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal frequency As Single, ByVal dfdt As Single)

        MyBase.New(parent, frequencyDefinition)

        m_frequency = frequency
        m_dfdt = dfdt

        m_frequencyAssigned = Not Single.IsNaN(frequency)
        m_dfdtAssigned = Not Single.IsNaN(dfdt)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal unscaledFrequency As Int16, ByVal unscaledDfDt As Int16)
    Protected Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal unscaledFrequency As Int16, ByVal unscaledDfDt As Int16)

        MyClass.New(parent, frequencyDefinition, _
            Convert.ToSingle(unscaledFrequency / frequencyDefinition.ScalingFactor + frequencyDefinition.Offset), _
            Convert.ToSingle(unscaledDfDt / frequencyDefinition.DfDtScalingFactor + frequencyDefinition.DfDtOffset))

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)
    Protected Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(parent, frequencyDefinition)
        ParseBinaryImage(Nothing, binaryImage, startIndex)

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal frequencyValue As IFrequencyValue)
    Protected Sub New(ByVal parent As IDataCell, ByVal frequencyDefinition As IFrequencyDefinition, ByVal frequencyValue As IFrequencyValue)

        MyClass.New(parent, frequencyDefinition, frequencyValue.Frequency, frequencyValue.DfDt)

    End Sub

    Public Overridable Property Frequency() As Single Implements IFrequencyValue.Frequency
        Get
            Return m_frequency
        End Get
        Set(ByVal value As Single)
            m_frequency = value
            m_frequencyAssigned = True
        End Set
    End Property

    Public Overridable Property DfDt() As Single Implements IFrequencyValue.DfDt
        Get
            Return m_dfdt
        End Get
        Set(ByVal value As Single)
            m_dfdt = value
            m_dfdtAssigned = True
        End Set
    End Property

    Public Overridable Property UnscaledFrequency() As Int16 Implements IFrequencyValue.UnscaledFrequency
        Get
            Try
                With Definition
                    Return Convert.ToInt16((m_frequency - .Offset) * .ScalingFactor)
                End With
            Catch ex As OverflowException
                Return Int16.MinValue
            End Try
        End Get
        Set(ByVal value As Int16)
            With Definition
                m_frequency = value / .ScalingFactor + .Offset
            End With
            m_frequencyAssigned = True
        End Set
    End Property

    Public Overridable Property UnscaledDfDt() As Int16 Implements IFrequencyValue.UnscaledDfDt
        Get
            Try
                With Definition
                    Return Convert.ToInt16((m_dfdt - .DfDtOffset) * .DfDtScalingFactor)
                End With
            Catch ex As OverflowException
                Return Int16.MinValue
            End Try
        End Get
        Set(ByVal value As Int16)
            With Definition
                m_dfdt = value / .DfDtScalingFactor + .DfDtOffset
            End With
            m_dfdtAssigned = True
        End Set
    End Property

    Default Public Overrides Property CompositeValue(ByVal index As Integer) As Single
        Get
            Select Case index
                Case CompositeValueType.Frequency
                    Return m_frequency
                Case CompositeValueType.DfDt
                    Return m_dfdt
                Case Else
                    Throw New IndexOutOfRangeException("Specified frequency value composite index, " & index & ", is out of range - there are only two composite values for a frequency value: frequency (0) and df/dt (1)")
            End Select
        End Get
        Set(ByVal value As Single)
            Select Case index
                Case CompositeValueType.Frequency
                    m_frequency = value
                    m_frequencyAssigned = True
                Case CompositeValueType.DfDt
                    m_dfdt = value
                    m_dfdtAssigned = True
                Case Else
                    Throw New IndexOutOfRangeException("Specified frequency value composite index, " & index & ", is out of range - there are only two composite values for a frequency value: frequency (0) and df/dt (1)")
            End Select
        End Set
    End Property

    Public Overrides ReadOnly Property CompositeValueCount() As Integer
        Get
            Return 2
        End Get
    End Property

    Public Overrides ReadOnly Property IsEmpty() As Boolean
        Get
            Return (Not m_frequencyAssigned OrElse Not m_dfdtAssigned)
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyLength() As UInt16
        Get
            If DataFormat = PhasorProtocols.DataFormat.FixedInteger Then
                Return 4
            Else
                Return 8
            End If
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyImage() As Byte()
        Get
            Dim buffer As Byte() = CreateArray(Of Byte)(BodyLength)

            If DataFormat = PhasorProtocols.DataFormat.FixedInteger Then
                EndianOrder.BigEndian.CopyBytes(UnscaledFrequency, buffer, 0)
                EndianOrder.BigEndian.CopyBytes(UnscaledDfDt, buffer, 2)
            Else
                EndianOrder.BigEndian.CopyBytes(m_frequency, buffer, 0)
                EndianOrder.BigEndian.CopyBytes(m_dfdt, buffer, 4)
            End If

            Return buffer
        End Get
    End Property

    Protected Overrides Sub ParseBodyImage(ByVal state As IChannelParsingState, ByVal binaryImage() As Byte, ByVal startIndex As Integer)

        If DataFormat = PhasorProtocols.DataFormat.FixedInteger Then
            UnscaledFrequency = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex)
            UnscaledDfDt = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 2)
        Else
            m_frequency = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex)
            m_dfdt = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex + 4)

            m_frequencyAssigned = True
            m_dfdtAssigned = True
        End If

    End Sub

    Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

        MyBase.GetObjectData(info, context)

        ' Serialize frequency value
        info.AddValue("frequency", m_frequency)
        info.AddValue("dfdt", m_dfdt)

    End Sub

    Public Overrides ReadOnly Property Attributes() As Dictionary(Of String, String)
        Get
            Dim baseAttributes As Dictionary(Of String, String) = MyBase.Attributes

            baseAttributes.Add("Frequency Value", Frequency)
            baseAttributes.Add("df/dt Value", DfDt)
            baseAttributes.Add("Unscaled Frequency Value", UnscaledFrequency)
            baseAttributes.Add("Unscaled df/dt Value", UnscaledDfDt)

            Return baseAttributes
        End Get
    End Property

End Class

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/FrequencyValueBase.cs
//*******************************************************************************************************
//  FrequencyValueBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  11/12/2004 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using PCS;
using PCS.Interop;

namespace PCS.PhasorProtocols
{
    /// <summary>
    /// Represents the common implementation of the protocol independent representation of a frequency and df/dt value.
    /// </summary>
    [Serializable()]
    public abstract class FrequencyValueBase : ChannelValueBase<IFrequencyDefinition>, IFrequencyValue
    {
        #region [ Members ]

        // Fields
        private double m_frequency;
        private double m_dfdt;
        private bool m_frequencyAssigned;
        private bool m_dfdtAssigned;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="FrequencyValueBase"/>.
        /// </summary>
        protected FrequencyValueBase()
        {
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyValueBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected FrequencyValueBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize frequency value
            m_frequency = info.GetDouble("frequency");
            m_dfdt = info.GetDouble("dfdt");

            m_frequencyAssigned = true;
            m_dfdtAssigned = true;
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyValueBase"/> from the specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IDataCell"/> parent of this <see cref="FrequencyValueBase"/>.</param>
        /// <param name="frequencyDefinition">The <see cref="IFrequencyDefinition"/> associated with this <see cref="FrequencyValueBase"/>.</param>
        /// <param name="frequency">The floating point value that represents this <see cref="FrequencyValueBase"/>.</param>
        /// <param name="dfdt">The floating point value that represents the change in this <see cref="FrequencyValueBase"/> over time.</param>
        protected FrequencyValueBase(IDataCell parent, IFrequencyDefinition frequencyDefinition, double frequency, double dfdt)
            : base(parent, frequencyDefinition)
        {
            m_frequency = frequency;
            m_dfdt = dfdt;

            m_frequencyAssigned = !double.IsNaN(frequency);
            m_dfdtAssigned = !double.IsNaN(dfdt);
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets or sets the floating point value that represents this <see cref="FrequencyValueBase"/>.
        /// </summary>
        public virtual double Frequency
        {
            get
            {
                return m_frequency;
            }
            set
            {
                m_frequency = value;
                m_frequencyAssigned = true;
            }
        }

        /// <summary>
        /// Gets or sets the floating point value that represents the change in this <see cref="FrequencyValueBase"/> over time.
        /// </summary>
        public virtual double DfDt
        {
            get
            {
                return m_dfdt;
            }
            set
            {
                m_dfdt = value;
                m_dfdtAssigned = true;
            }
        }

        /// <summary>
        /// Gets or sets the unscaled integer representation of this <see cref="FrequencyValueBase"/>.
        /// </summary>
        public virtual int UnscaledFrequency
        {
            get
            {
                unchecked
                {
                    return (int)((m_frequency - Definition.Offset) * Definition.ScalingValue);
                }
            }
            set
            {
                m_frequency = value / (double)Definition.ScalingValue + Definition.Offset;
                m_frequencyAssigned = true;
            }
        }

        /// <summary>
        /// Gets or sets the unscaled integer representation of the change in this <see cref="FrequencyValueBase"/> over time.
        /// </summary>
        public virtual int UnscaledDfDt
        {
            get
            {
                unchecked
                {
                    return (int)((m_dfdt - Definition.DfDtOffset) * Definition.DfDtScalingValue);
                }
            }
            set
            {
                m_dfdt = value / (double)Definition.DfDtScalingValue + Definition.DfDtOffset;
                m_dfdtAssigned = true;
            }
        }

        /// <summary>
        /// Gets boolean value that determines if none of the composite values of <see cref="FrequencyValueBase"/> have been assigned a value.
        /// </summary>
        /// <returns>True, if no composite values have been assigned a value; otherwise, false.</returns>
        public override bool IsEmpty
        {
            get
            {
                return (!m_frequencyAssigned || !m_dfdtAssigned);
            }
        }

        /// <summary>
        /// Gets the composite values of this <see cref="FrequencyValueBase"/>.
        /// </summary>
        public override double[] CompositeValues
        {
            get
            {
                return new double[] { m_frequency, m_dfdt };
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="BodyImage"/>.
        /// </summary>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override int BodyLength
        {
            get
            {
                if (DataFormat == PhasorProtocols.DataFormat.FixedInteger)
                    return 4;
                else
                    return 8;
            }
        }

        /// <summary>
        /// Gets the binary body image of the <see cref="FrequencyValueBase"/> object.
        /// </summary>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override byte[] BodyImage
        {
            get
            {
                byte[] buffer = new byte[BodyLength];

                // Had to make a descision on usage versus typical protocol implementation when
                // exposing values as double / int when protocols typically use float / short for
                // transmission. Exposing values as double / int makes class more versatile by
                // allowing future protocol implementations to support higher resolution values
                // simply by overriding BodyLength, BodyImage and ParseBodyImage. Exposing class
                // values as double / int runs the risk of providing values that are outside the
                // data type limitations, hence the unchecked section below. However, risk should
                // be low in typical usage scenarios since values being transmitted via a generated
                // image were likely parsed previously from a binary image with the same constraints.
                unchecked
                {
                    if (DataFormat == PhasorProtocols.DataFormat.FixedInteger)
                    {
                        EndianOrder.BigEndian.CopyBytes((short)UnscaledFrequency, buffer, 0);
                        EndianOrder.BigEndian.CopyBytes((short)UnscaledDfDt, buffer, 2);
                    }
                    else
                    {
                        EndianOrder.BigEndian.CopyBytes((float)m_frequency, buffer, 0);
                        EndianOrder.BigEndian.CopyBytes((float)m_dfdt, buffer, 4);
                    }
                }

                return buffer;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="FrequencyValueBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Frequency Value", Frequency.ToString());
                baseAttributes.Add("df/dt Value", DfDt.ToString());
                baseAttributes.Add("Unscaled Frequency Value", UnscaledFrequency.ToString());
                baseAttributes.Add("Unscaled df/dt Value", UnscaledDfDt.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Parses the binary body image.
        /// </summary>
        /// <param name="binaryImage">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="binaryImage"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="binaryImage"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        /// <remarks>
        /// The base implementation assumes fixed integer values are represented as 16-bit signed
        /// integers and floating point values are represented as 32-bit single-precision floating-point
        /// values (i.e., short and float data types respectively).
        /// </remarks>
        protected override int ParseBodyImage(byte[] binaryImage, int startIndex, int length)
        {
            // Length is validated at a frame level well in advance so that low level parsing routines do not have
            // to re-validate that enough length is available to parse needed information as an optimization...

            if (DataFormat == PhasorProtocols.DataFormat.FixedInteger)
            {
                UnscaledFrequency = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex);
                UnscaledDfDt = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 2);
                
                return 4;
            }
            else
            {
                m_frequency = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex);
                m_dfdt = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex + 4);

                m_frequencyAssigned = true;
                m_dfdtAssigned = true;

                return 8;
            }
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize frequency value
            info.AddValue("frequency", m_frequency);
            info.AddValue("dfdt", m_dfdt);
        }

        #endregion
    }
}